homeassistant/components/openevse/__init__.py

```diff
@@ -1 +1,51 @@
 """The openevse component."""
+import logging
+
+import openevsewifi
+from openevsewifi import InvalidAuthentication
+from requests import RequestException
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN, PLATFORMS
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Set up is called when Home Assistant is loading our component."""
+    config_entry.add_update_listener(update_listener)
+    await hass.config_entries.async_forward_entry_setup(config_entry, PLATFORMS)
+
+    return True
+
+
+async def update_listener(hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+    """Update listener."""
+
+    _LOGGER.debug("Attempting to reload entities from the %s integration", DOMAIN)
+
+    if config_entry.data == config_entry.options:
+        _LOGGER.debug("No changes detected not reloading entities")
+        return
+
+    new_data = config_entry.options.copy()
+
+    hass.config_entries.async_update_entry(
+        entry=config_entry,
+        data=new_data,
+    )
+
+    await hass.config_entries.async_reload(config_entry.entry_id)
+
+
+def test_connection(host: str, username: str, password: str) -> bool:
+    """Test connection to charger."""
+    charger = openevsewifi.Charger(host, username=username, password=password)
+    try:
+        charger.status
+    except (RequestException, InvalidAuthentication) as e:
```

Comment:
I'm having trouble mocking the library's `InvalidAuthentication` exception, any pointers?

Reply:
This is how I did it in the [Nanoleaf config flow tests](https://github.com/home-assistant/core/blob/18c03e2f8db172d7f6386a59f3364705b6ebb87b/tests/components/nanoleaf/test_config_flow.py#L114):

```python3
with patch(
    "homeassistant.components.openevse.__init__.openevsewifi.Charger",
) as mock_charger:
    charger = mock_charger.return_value
    charger.status.side_effect = InvalidAuthentication("InvalidAuthentication")
```

Reply:
Unfortunately I can't help you much with this. I spent a few hours looking for this Nanoleaf solution last week.

Reply:
Oh, I need to mock requests as well that's what it is.